File: frontend/lib/features/experiment4_food/services/food_service.dart
import '../../../core/network/api_client.dart';
import '../models/address.dart';
import '../models/cart.dart';
import '../models/food_item.dart';
import '../models/order.dart';
import '../models/restaurant.dart';

class RestaurantMenu {
  final Restaurant restaurant;
  final List<FoodItem> items;

  const RestaurantMenu({required this.restaurant, required this.items});

  factory RestaurantMenu.fromJson(Map<String, dynamic> json) => RestaurantMenu(
        restaurant: Restaurant.fromJson(
            (json['restaurant'] as Map<String, dynamic>?) ?? const {}),
        items: ((json['items'] as List?) ?? const [])
            .map((e) => FoodItem.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class FoodService {
  final ApiClient _api = ApiClient.instance;

  Future<List<Restaurant>> listRestaurants({
    String? search,
    String? cuisine,
    String? city,
    double? minRating,
    String sort = 'name',
  }) async {
    final data = await _api.get('/api/restaurants', query: {
      if (search != null && search.trim().isNotEmpty) 'search': search.trim(),
      if (cuisine != null && cuisine.isNotEmpty) 'cuisine': cuisine,
      if (city != null && city.isNotEmpty) 'city': city,
      if (minRating != null) 'min_rating': minRating,
      'sort': sort,
    });
    return (data as List)
        .map((e) => Restaurant.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<Restaurant> getRestaurant(String restaurantId) async {
    final data = await _api.get('/api/restaurants/$restaurantId');
    return Restaurant.fromJson(data as Map<String, dynamic>);
  }

  Future<RestaurantMenu> getMenu(String restaurantId,
      {String? category}) async {
    final data = await _api.get('/api/restaurants/$restaurantId/menu', query: {
      if (category != null && category.isNotEmpty) 'category': category,
    });
    return RestaurantMenu.fromJson(data as Map<String, dynamic>);
  }

  Future<List<Map<String, dynamic>>> listReviews(String restaurantId) async {
    final data = await _api.get('/api/restaurants/$restaurantId/reviews');
    return (data as List)
        .map((e) => e as Map<String, dynamic>)
        .toList();
  }

  Future<Cart> getCart() async =>
      Cart.fromJson(await _api.get('/api/cart') as Map<String, dynamic>);

  Future<Cart> addToCart({
    required String restaurantId,
    required String foodItemId,
    int quantity = 1,
  }) async {
    final data = await _api.post('/api/cart', body: {
      'restaurant_id': restaurantId,
      'food_item_id': foodItemId,
      'quantity': quantity,
    });
    return Cart.fromJson(data as Map<String, dynamic>);
  }

  Future<Cart> updateCartItem(String foodItemId, int quantity) async {
    final data = await _api
        .put('/api/cart/$foodItemId', body: {'quantity': quantity});
    return Cart.fromJson(data as Map<String, dynamic>);
  }

  Future<Cart> removeCartItem(String foodItemId) async {
    final data = await _api.delete('/api/cart/$foodItemId');
    return Cart.fromJson(data as Map<String, dynamic>);
  }

  Future<List<DeliveryAddress>> listAddresses() async {
    final data = await _api.get('/api/addresses');
    return (data as List)
        .map((e) => DeliveryAddress.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<DeliveryAddress> addAddress({
    required String label,
    required String fullName,
    required String phone,
    required String addressLine,
    required String city,
    required String state,
    required String postalCode,
  }) async {
    final data = await _api.post('/api/addresses', body: {
      'label': label,
      'full_name': fullName,
      'phone': phone,
      'address_line': addressLine,
      'city': city,
      'state': state,
      'postal_code': postalCode,
    });
    return DeliveryAddress.fromJson(data as Map<String, dynamic>);
  }

  Future<void> deleteAddress(String addressId) async {
    await _api.delete('/api/addresses/$addressId');
  }

  Future<Order> placeOrder({
    required String restaurantId,
    required List<Map<String, dynamic>> items,
    required String addressId,
    required String paymentMethod,
  }) async {
    final data = await _api.post('/api/orders', body: {
      'restaurant_id': restaurantId,
      'items': items,
      'address_id': addressId,
      'payment_method': paymentMethod,
    });
    return Order.fromJson(data as Map<String, dynamic>);
  }

  Future<List<Order>> listOrders() async {
    final data = await _api.get('/api/orders');
    return (data as List)
        .map((e) => Order.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<Order> getOrder(String orderId) async {
    final data = await _api.get('/api/orders/$orderId');
    return Order.fromJson(data as Map<String, dynamic>);
  }

  Future<Restaurant> createRestaurant({
    required String name,
    required String cuisine,
    required String city,
    required String address,
    String phone = '',
    String description = '',
    String imageUrl = '',
    int deliveryEstimate = 30,
    double deliveryFee = 0,
  }) async {
    final data = await _api.post('/api/admin/restaurants', body: {
      'name': name,
      'cuisine': cuisine,
      'city': city,
      'address': address,
      'phone': phone,
      'description': description,
      'image_url': imageUrl,
      'delivery_estimate': deliveryEstimate,
      'delivery_fee': deliveryFee,
    });
    return Restaurant.fromJson(data as Map<String, dynamic>);
  }

  Future<Restaurant> updateRestaurant(
    String restaurantId, {
    required String name,
    required String cuisine,
    required String city,
    required String address,
    String phone = '',
    String description = '',
    String imageUrl = '',
    int deliveryEstimate = 30,
    double deliveryFee = 0,
  }) async {
    final data = await _api.put('/api/admin/restaurants/$restaurantId', body: {
      'name': name,
      'cuisine': cuisine,
      'city': city,
      'address': address,
      'phone': phone,
      'description': description,
      'image_url': imageUrl,
      'delivery_estimate': deliveryEstimate,
      'delivery_fee': deliveryFee,
    });
    return Restaurant.fromJson(data as Map<String, dynamic>);
  }

  Future<void> deleteRestaurant(String restaurantId) async {
    await _api.delete('/api/admin/restaurants/$restaurantId');
  }

  Future<FoodItem> createFoodItem({
    required String restaurantId,
    required String name,
    String description = '',
    required double price,
    required String category,
    String imageUrl = '',
    bool isVegetarian = false,
    bool isAvailable = true,
  }) async {
    final data = await _api.post(
        '/api/admin/restaurants/$restaurantId/food',
        body: {
          'restaurant_id': restaurantId,
          'name': name,
          'description': description,
          'price': price,
          'category': category,
          'image_url': imageUrl,
          'is_vegetarian': isVegetarian,
          'is_available': isAvailable,
        });
    return FoodItem.fromJson(data as Map<String, dynamic>);
  }

  Future<FoodItem> updateFoodItem(
    String foodItemId, {
    required String restaurantId,
    required String name,
    String description = '',
    required double price,
    required String category,
    String imageUrl = '',
    bool isVegetarian = false,
    bool isAvailable = true,
  }) async {
    final data = await _api.put(
        '/api/admin/restaurants/food/$foodItemId',
        body: {
          'restaurant_id': restaurantId,
          'name': name,
          'description': description,
          'price': price,
          'category': category,
          'image_url': imageUrl,
          'is_vegetarian': isVegetarian,
          'is_available': isAvailable,
        });
    return FoodItem.fromJson(data as Map<String, dynamic>);
  }

  Future<void> deleteFoodItem(String foodItemId) async {
    await _api.delete('/api/admin/restaurants/food/$foodItemId');
  }
}

File: backend/app/services/food_service.py
from ..repositories.food_repo import (
    AddressRepository, CartRepository, FoodItemRepository, OrderRepository,
    RestaurantRepository, ReviewRepository,
)
from ..utils.errors import BadRequestException, ConflictException, NotFoundException
from ..utils.helpers import parse_object_id, utc_now_iso

ORDER_STATUSES = ("PLACED", "CONFIRMED", "PREPARING", "OUT_FOR_DELIVERY", "DELIVERED", "CANCELLED")


class FoodService:
    def __init__(self):
        self.restaurants = RestaurantRepository()
        self.food = FoodItemRepository()
        self.carts = CartRepository()
        self.addresses = AddressRepository()
        self.orders = OrderRepository()
        self.reviews = ReviewRepository()

    # Restaurants -----------------------------------------------------------
    async def list_restaurants(self, search, cuisine, city, min_rating, sort, skip, limit) -> list[dict]:
        restaurants = await self.restaurants.list_all(
            search=search, cuisine=cuisine, city=city, min_rating=min_rating,
            sort=sort, skip=skip, limit=limit,
        )
        for r in restaurants:
            r["rating"] = r.get("rating", 0.0)
            r["rating_count"] = r.get("rating_count", 0)
        return restaurants

    async def get_restaurant(self, restaurant_id: str) -> dict:
        oid = parse_object_id(restaurant_id, "restaurant id")
        restaurant = await self.restaurants.find_by_id(oid)
        if restaurant is None:
            raise NotFoundException("Restaurant not found", "RESTAURANT_NOT_FOUND")
        restaurant["rating"] = restaurant.get("rating", 0.0)
        restaurant["rating_count"] = restaurant.get("rating_count", 0)
        return restaurant

    async def restaurant_menu(self, restaurant_id: str, category: str | None) -> dict:
        oid = parse_object_id(restaurant_id, "restaurant id")
        restaurant = await self.restaurants.find_by_id(oid)
        if restaurant is None:
            raise NotFoundException("Restaurant not found", "RESTAURANT_NOT_FOUND")
        items = await self.food.list_by_restaurant(str(oid), category)
        return {"restaurant": restaurant, "items": items}

    # Cart ----------------------------------------------------------------
    async def get_cart(self, user_id: str) -> dict:
        cart = await self.carts.get_cart(user_id)
        if cart is None:
            return {"restaurant_id": None, "items": [], "total": 0.0}
        items = []
        for item in cart.get("items", []):
            food = await self.food.find_by_id(parse_object_id(item["food_item_id"], "food item id"))
            if food:
                items.append({**item, "food_item": food, "subtotal": round(food["price"] * item["quantity"], 2)})
        total = round(sum(i["subtotal"] for i in items), 2)
        return {"restaurant_id": cart.get("restaurant_id"), "items": items, "total": total}

    async def add_to_cart(self, user_id: str, restaurant_id: str, food_item_id: str, quantity: int) -> dict:
        rid = parse_object_id(restaurant_id, "restaurant id")
        fid = parse_object_id(food_item_id, "food item id")
        restaurant = await self.restaurants.find_by_id(rid)
        if restaurant is None:
            raise NotFoundException("Restaurant not found", "RESTAURANT_NOT_FOUND")
        food = await self.food.find_by_id(fid)
        if food is None or food.get("restaurant_id") != str(rid):
            raise NotFoundException("Food item not found in this restaurant", "FOOD_NOT_FOUND")
        if not food.get("is_available", True):
            raise BadRequestException("This item is currently unavailable", "ITEM_UNAVAILABLE")

        cart = await self.carts.get_cart(user_id)
        items = []
        if cart and cart.get("restaurant_id") and cart["restaurant_id"] != str(rid):
            items = []
        elif cart:
            items = cart.get("items", [])

        found = False
        for item in items:
            if item["food_item_id"] == str(fid):
                item["quantity"] = min(item["quantity"] + quantity, 50)
                found = True
                break
        if not found:
            items.append({"food_item_id": str(fid), "quantity": quantity})

        total = 0.0
        for item in items:
            f = await self.food.find_by_id(parse_object_id(item["food_item_id"], "food item id"))
            if f:
                total += f["price"] * item["quantity"]
        cart = await self.carts.upsert_cart(user_id, items, str(rid), round(total, 2))
        return await self.get_cart(user_id)

    async def update_cart_item(self, user_id: str, food_item_id: str, quantity: int) -> dict:
        fid = parse_object_id(food_item_id, "food item id")
        cart = await self.carts.get_cart(user_id)
        if cart is None or not cart.get("items"):
            raise NotFoundException("Cart is empty", "CART_EMPTY")
        items = cart["items"]
        found = False
        for item in items:
            if item["food_item_id"] == str(fid):
                item["quantity"] = quantity
                found = True
        if not found:
            raise NotFoundException("Item not in cart", "ITEM_NOT_IN_CART")
        total = 0.0
        for item in items:
            f = await self.food.find_by_id(parse_object_id(item["food_item_id"], "food item id"))
            if f:
                total += f["price"] * item["quantity"]
        await self.carts.upsert_cart(user_id, items, cart.get("restaurant_id"), round(total, 2))
        return await self.get_cart(user_id)

    async def remove_cart_item(self, user_id: str, food_item_id: str) -> dict:
        fid = parse_object_id(food_item_id, "food item id")
        cart = await self.carts.get_cart(user_id)
        if cart is None or not cart.get("items"):
            raise NotFoundException("Cart is empty", "CART_EMPTY")
        items = [i for i in cart["items"] if i["food_item_id"] != str(fid)]
        if len(items) == len(cart["items"]):
            raise NotFoundException("Item not in cart", "ITEM_NOT_IN_CART")
        total = 0.0
        for item in items:
            f = await self.food.find_by_id(parse_object_id(item["food_item_id"], "food item id"))
            if f:
                total += f["price"] * item["quantity"]
        if not items:
            await self.carts.clear(user_id)
            return {"restaurant_id": None, "items": [], "total": 0.0}
        await self.carts.upsert_cart(user_id, items, cart.get("restaurant_id"), round(total, 2))
        return await self.get_cart(user_id)

    # Addresses -----------------------------------------------------------
    async def list_addresses(self, user_id: str) -> list[dict]:
        return await self.addresses.list_for_user(user_id)

    async def add_address(self, user_id: str, address: dict) -> dict:
        return await self.addresses.create(user_id, address)

    async def delete_address(self, user_id: str, address_id: str) -> None:
        oid = parse_object_id(address_id, "address id")
        address = await self.addresses.find_for_user(oid, user_id)
        if address is None:
            raise NotFoundException("Address not found", "ADDRESS_NOT_FOUND")
        await self.addresses.delete(oid)

    # Orders --------------------------------------------------------------
    async def place_order(self, user_id: str, restaurant_id: str, items: list[dict], address_id: str,
                          payment_method: str) -> dict:
        rid = parse_object_id(restaurant_id, "restaurant id")
        aid = parse_object_id(address_id, "address id")
        restaurant = await self.restaurants.find_by_id(rid)
        if restaurant is None:
            raise NotFoundException("Restaurant not found", "RESTAURANT_NOT_FOUND")
        address = await self.addresses.find_for_user(aid, user_id)
        if address is None:
            raise NotFoundException("Address not found", "ADDRESS_NOT_FOUND")

        order_items = []
        total = 0.0
        for item in items:
            fid = parse_object_id(item["food_item_id"], "food item id")
            food = await self.food.find_by_id(fid)
            if food is None or food.get("restaurant_id") != str(rid):
                raise NotFoundException("Food item not found", "FOOD_NOT_FOUND")
            if not food.get("is_available", True):
                raise BadRequestException(f"{food['name']} is unavailable", "ITEM_UNAVAILABLE")
            subtotal = round(food["price"] * item["quantity"], 2)
            total += subtotal
            order_items.append({
                "food_item_id": str(fid),
                "name": food["name"],
                "price": food["price"],
                "quantity": item["quantity"],
                "subtotal": subtotal,
            })
        total = round(total, 2)
        delivery_fee = restaurant.get("delivery_fee", 0)
        order = {
            "user_id": user_id,
            "restaurant_id": str(rid),
            "restaurant_name": restaurant["name"],
            "items": order_items,
            "subtotal": total,
            "delivery_fee": delivery_fee,
            "total": round(total + delivery_fee, 2),
            "address": {k: address[k] for k in ("label", "full_name", "phone", "address_line", "city", "state", "postal_code") if k in address},
            "payment_method": payment_method.upper() if payment_method else "CASH",
            "status": "PLACED",
            "status_history": [{"status": "PLACED", "at": utc_now_iso(), "note": "Order placed"}],
            "created_at": utc_now_iso(),
        }
        created = await self.orders.create(order)
        await self.carts.clear(user_id)
        return created

    async def list_orders(self, user_id: str, status: str | None) -> list[dict]:
        return await self.orders.list_for_user(user_id, status)

    async def get_order(self, user_id: str, order_id: str) -> dict:
        oid = parse_object_id(order_id, "order id")
        order = await self.orders.find_for_user(oid, user_id)
        if order is None:
            raise NotFoundException("Order not found", "ORDER_NOT_FOUND")
        return order

    async def update_order_status(self, order_id: str, status: str, note: str | None = None) -> dict:
        oid = parse_object_id(order_id, "order id")
        order = await self.orders.find_by_id(oid)
        if order is None:
            raise NotFoundException("Order not found", "ORDER_NOT_FOUND")
        if status not in ORDER_STATUSES:
            raise BadRequestException("Invalid order status", "INVALID_STATUS")
        if order["status"] == "CANCELLED" or order["status"] == "DELIVERED":
            raise ConflictException(f"Order already {order['status'].lower()}", "ORDER_CLOSED")
        history = order.get("status_history", [])
        history.append({"status": status, "at": utc_now_iso(), "note": note or ""})
        return await self.orders.update(oid, {"status": status, "status_history": history})

    # Reviews -------------------------------------------------------------
    async def add_review(self, user_id: str, restaurant_id: str, rating: int, comment: str) -> dict:
        rid = parse_object_id(restaurant_id, "restaurant id")
        restaurant = await self.restaurants.find_by_id(rid)
        if restaurant is None:
            raise NotFoundException("Restaurant not found", "RESTAURANT_NOT_FOUND")
        review = {
            "user_id": user_id,
            "restaurant_id": str(rid),
            "rating": int(rating),
            "comment": comment.strip()[:500],
            "created_at": utc_now_iso(),
        }
        created = await self.reviews.create(review)
        avg = await self.reviews.avg_rating(str(rid))
        count = await self.reviews.for_restaurant(str(rid))
        await self.restaurants.update(rid, {"rating": avg or 0.0, "rating_count": len(count)})
        return created

    async def list_reviews(self, restaurant_id: str) -> list[dict]:
        oid = parse_object_id(restaurant_id, "restaurant id")
        reviews = await self.reviews.for_restaurant(str(oid))
        from ..repositories.users_repo import UserRepository
        user_repo = UserRepository()
        for r in reviews:
            try:
                user = await user_repo.find_by_id(parse_object_id(r["user_id"], "user id"))
                r["reviewer_name"] = user["name"] if user else "Anonymous"
            except Exception:
                r["reviewer_name"] = "Anonymous"
        return reviews